File: LittleProjects/Javascript/NodeCRUD/core/repository/repository.js
const pool = require('../datasource/pgsource');

class Repository {
    constructor(schema, table, entity) {
        this.schema = schema;
        this.table = table;
        this.entity = entity;
    }

    save(entity) {
        return pool.query(`insert into ${this.schema}.${this.table}(${this.entity.fields().slice(1).map(field => `"${field}"`).join()}) values(${this.entity.fields().slice(1).map((field, index) => `$${index + 1}`)}) returning "${this.entity.fields()[0]}"`,
            entity.values().slice(1))
            .then(result => {
                entity.setId(result.rows[0][this.entity.fields()[0]]);
                return entity;
            })
            .catch(err => JSON.stringify(err));
    }

    get(id) {
        return pool.query(`select ${this.entity.fields().join()} from ${this.schema}.${this.table} where ${this.entity.fields()[0]} = $1`, [id])
            .then(result => this.entity.apply(result.rows))
            .catch(err => JSON.stringify(err));
    }

    remove(id) {
        return pool.query(`delete from ${this.schema}.${this.table} where ${this.entity.fields()[0]} = $1`, [id])
            .then(_ =>  JSON.stringify({"status": "ok"}))
            .catch(err =>  JSON.stringify(err))
    }

    update(entity) {
        console.log(`update ${this.schema}.${this.table} set ${this.entity.fields().slice(1).map((field, index) => `${field} = $${index + 2}`)} where ${this.entity.fields()[0]} = $1`);
        return pool.query(`update ${this.schema}.${this.table} set ${this.entity.fields().slice(1).map((field, index) => `${field} = $${index + 2}`)} where ${this.entity.fields()[0]} = $1`,
            entity.values())
            .then(_ =>  JSON.stringify({"status": "ok"}))
            .catch(err => JSON.stringify(err))
    }
}

module.exports = Repository;

File: LittleProjects/Python/TornadoPostgresCrud/app/repository.py
from app.common import Singleton, log
from app.datasource import transaction


class Repository(object, metaclass=Singleton):
    def __init__(self):
        #  bad practice, but in this case - ok
        with transaction() as session:
            session.execute('create schema if not exists crud')
            session.execute('create sequence if not exists crud.entity_seq')
            session.execute("create table if not exists crud.entity(id numeric(38) primary key default nextval('crud.entity_seq'), value varchar(255))")

    @log
    def get_all(self):
        with transaction() as session:
            session.execute('select id, value from crud.entity')
            return session.fetchall()

    @log
    def get_by_id(self, id):
        with transaction() as session:
            session.execute('select id, value from crud.entity where id = %s', (id,))
            return session.fetchone()

    @log
    def save(self, entity):
        with transaction() as session:
            session.execute('insert into crud.entity(value) values (%s) returning id', (entity.value,))
            next_id = session.fetchone()
            entity.id = int(next_id)
            return entity

    @log
    def delete_by_id(self, id):
        with transaction() as session:
            session.execute('delete from crud.entity where id = %s', (id, ))
            return 'OK'

    @log
    def update(self, entity):
        with transaction() as session:
            session.execute('update crud.entity set value = %s where id = %s', (entity.value, entity.id))
            return entity
